### PR TITLE
nixos/hans: init

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

Add module for Hans IP over ICMP tunnel. Mostly copypaste of iodine module.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/networking/hans.nix [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        serviceConfig.ExecStart = "${pkgs.hans}/bin/hans -f -u ${hansUser} ${cfg.server.extraConfig} -s ${cfg.server.ip} ${optionalString cfg.server.systemPings "-r"}";
+      };
+    };
+
+    users.extraUsers = singleton {
+      name = hansUser;
+      uid = config.ids.uids.hans;
+      description = "Hans daemon user";
+    };
+    users.extraGroups.hans.gid = config.ids.gids.hans;
```
Comment (Mic92) [inline nixos/modules/services/networking/hans.nix]: Fixed uid/gids should not be required since hans has no persistent files.

**Comment by gnidorah** [comment]
@Mic92 Fixed, thank you!

File: nixos/modules/services/networking/iodine.nix [hunk 2]
```diff
@@ -124,7 +144,15 @@ in
...
         wantedBy = [ "multi-user.target" ];
-        serviceConfig.ExecStart = "${pkgs.iodine}/bin/iodined -f -u ${iodinedUser} ${cfg.server.extraConfig} ${cfg.server.ip} ${cfg.server.domain}";
+        preStart = ''
+          cat > /run/iodined/script << EOF
+            #!/bin/sh
+            ${pkgs.iodine}/bin/iodined -f -u ${iodinedUser} ${cfg.server.extraConfig} ${optionalString (cfg.passwordFile != "") "-P $(cat \"${cfg.passwordFile}\")"} ${cfg.server.ip} ${cfg.server.domain}
+          EOF
+          chmod 700 /run/iodined/script
+        '';
+        script = "/run/iodined/script";
```
Comment (Mic92) [inline nixos/modules/services/networking/iodine.nix]: You can put the same script, you wrote into `/run/iodined/script` here as well.

File: nixos/modules/services/networking/iodine.nix [hunk 3]
```diff
@@ -108,10 +120,18 @@ in
...
         after = [ "network.target" ];
         wantedBy = [ "multi-user.target" ];
+        preStart = ''
+          cat > /run/iodine/${name}-script << EOF
+            #!/bin/sh
+            ${pkgs.iodine}/bin/iodine -f -u ${iodinedUser} ${cfg.extraConfig} ${optionalString (cfg.passwordFile != "") "-P $(cat \"${cfg.passwordFile}\")"} ${cfg.relay} ${cfg.server}
+          EOF
+          chmod 700 /run/iodine/${name}-script
+        '';
+        script = "/run/iodine/${name}-script";
```
Comment (Mic92) [inline nixos/modules/services/networking/iodine.nix]: same here.

File: nixos/modules/services/networking/hans.nix [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        preStart = ''
+          cat > /run/hans/script << EOF
+            #!/bin/sh
+            ${pkgs.hans}/bin/hans -f -u ${hansUser} ${cfg.server.extraConfig} -s ${cfg.server.ip} ${optionalString cfg.server.systemPings "-r"} ${optionalString (cfg.passwordFile != "") "-p $(cat \"${cfg.passwordFile}\")"}
+          EOF
+          chmod 700 /run/hans/script
+        '';
+        script = "/run/hans/script";
```
Comment (Mic92) [inline nixos/modules/services/networking/hans.nix]: same here.

File: nixos/modules/services/networking/hans.nix [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+      {
+        description = "hans client - ${name}";
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        preStart = ''
+          cat > /run/hans/${name}-script << EOF
+            #!/bin/sh
+            ${pkgs.hans}/bin/hans -f -u ${hansUser} ${cfg.extraConfig} -c ${cfg.server} ${optionalString (cfg.passwordFile != "") "-p $(cat \"${cfg.passwordFile}\")"} 
+          EOF
+          chmod 700 /run/hans/${name}-script
```
Comment (Mic92) [inline nixos/modules/services/networking/hans.nix]: same here

**Comment by gnidorah** [comment]
@Mic92 Thanks!

File: nixos/modules/services/networking/hans.nix [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+          default = "";
+          description = "The assigned ip range";
+          example = "198.51.100.0";
+        };
+
+        systemPings = mkOption {
+          type = types.bool;
+          default = false;
+          description = "Respond to ordinary pings";
+        };
```
Comment (fpletz) [inline nixos/modules/services/networking/hans.nix]: I don't think this option belongs in this module. We also have `networking.firewall.allowPing`.

Comment (gnidorah) [reply nixos/modules/services/networking/hans.nix]: @fpletz Renamed. This is pure hans option https://github.com/albertzak/hanstunnel#cant-connect

Comment (fpletz) [reply nixos/modules/services/networking/hans.nix]: Oh, I see. You're right. :+1:

**Comment by gnidorah** [comment]
@matthewbauer Thank you!

**Pull request merged**